### ISSUE TITLE
request does not appear on debug sidebar on certain exceptions (middleware or tween related?)

### ISSUE THREAD

**Issue by jvanasco** [Pylons/pyramid_debugtoolbar#237]
I can't track the cause of this down yet.

Occasionally I will have a "serious" error in my pyramid stack that generates a waitress error page:

```
Internal Server Error

The server encountered an unexpected internal server error

(generated by waitress)
```

I just noticed that when that happens, the request is not available to debugtoolbar's panel listing.  The page does have a traceback that is accessible via the log -- the error is just triggered in such a way that the particular component of history items remains unaware.

**Comment by bertjwregeer** [comment]
This means the tween didn't have a chance to capture the exception, and instead Waitress caught the exception instead.

Check where the debug toolbar tween is in the stack, and try and figure out what is throwing the exception.

**Comment by jvanasco** [comment]
Sorry, let me rephrase.  I've been staring at screens too long today...

I know how to troubleshoot this type of developer error (they admittedly happen often), however others probably won't.

There's no info in the documentation that debugtoolbar might not be able to catch all errors.

Although the toolbar's panel system doesn't have this error visible, there is a persistent object reference... somewhere.  

Also, I can't track down the exact cause for certain types of errors not getting caught by Pyramid and reproduce a test case yet, and I'd like to:

1. That reference (there is a functional debug page) might be somewhere that is accessible to the toolbar and can still populate it (somehow?).
2. From my experience with the toolbar's internals and how it handles history management + display, I'm not sure the max requests setting could be enforced on these pages.  If that is the case, this could lead to unrestrained process growth.  All those debug pages would be accessible and functional --  so it wouldn't be a memory leak -- but if the history limits don't apply to them, it would have the same effect.

**Comment by jvanasco** [comment]
Update: I managed to trigger this with an unhandled decoder hook in msgpack encoding for my cache.  I set the toolbar history to 5 requests, and kept hitting reload nonstop (100+ requests).  My process's memory jumped from 150MB to 900MB and the first request was still accessible (ie, not expired).

**Comment by jvanasco** [comment]
closing this. the newer way the toolbar wraps requests should fix it.
